### ISSUE TITLE
Error "Cannot create require path" with custom paths

### ISSUE THREAD

**Issue by danielpza** [TypeScriptToLua/TypeScriptToLua#694]
It seems tstl resolve method doesn't take into account the paths option in tsconfig.json, typescript doesn't detect any error, tstl does:

### Input:

```json
// tsconfig.json
{
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "lib",
    "baseUrl": "./",
    "paths": {
      "*": ["src/scripts/*"]
    }
  },
  "tstl": {
    "luaTarget": "5.1",
    "luaLibImport": "inline"
  }
}
```

```typescript
// src/main.ts
import { getPrefabCopy, PrefabLike } from "prefab-cache";
```

```typescript
/// src/scripts/prefab-cache.ts
export interface PrefabLike {}
export function getPrefabCopy(prefab: string) {/*...*/}
```

### Expected

```lua
-- lib/main.lua
require("prefab-cache")
```

### Output

```
src/modmain.ts:55:43 - error TSTL0: Cannot create require path. Module does not exist within --rootDir. TypeScript path: prefab-cache.

55 import { getPrefabCopy, PrefabLike } from "prefab-cache";
                                             ~~~~~~~~~~~~~~
```

**Comment by ark120202** [comment]
So you expect unresolved `require("prefab-cache")` as an output? Does `scripts` directory get some special behavior in your environment? So should it just ignore that it's unresolvable if there is some path mapping? But what if there is a valid resolved path and mapped one at the same time? Or what if mapped path is definitely not compatible with Lua, like `prefab-cache/foo`?

**Comment by danielpza** [comment]
> So you expect unresolved require("prefab-cache") as an output? 

yes

> Does scripts directory get some special behavior in your environment?

yes

> So should it just ignore that it's unresolvable if there is some path mapping? But what if there is a valid resolved path and mapped one at the same time? Or what if mapped path is definitely not compatible with Lua, like prefab-cache/foo?


Well, I would expect it to behave as typescript does, I guess these questions applies to typescript as well, but there are reasons for using path mapping

Also the error `tstl` is throwing is it cannot find the module, which exists, it just isn't using the typescript resolve algorithm.

https://www.typescriptlang.org/docs/handbook/module-resolution.html#path-mapping.

**Comment by ark120202** [comment]
TypeScript works in a different scope there. Since path resolution is a platform feature, TypeScript doesn't do that at all. It works only with type definitions, which can't be reliably mapped to real module files, and all TypeScript resolution options affect only that part. But tstl also has to implement host features, resolving the real imported file and formatting it to match Lua require logic.

I think the best way to solve it would be providing a new option that would mirror `LUA_PATH` in runtime. However that option should only change formatting of resolved paths, so `"./scripts/prefab-cache"` would first be resolved to absolute path and then resolved relatively to that option. Making `"prefab-cache"` resolve as `"./scripts/prefab-cache"` can be done with custom transformers (#591).

**Comment by danielpza** [comment]
@ark120202, so how would I go about fixing this error, and producing the correct path?

**Comment by ark120202** [comment]
The simplest hack I can think of would be making `src/scripts` the root directory and moving `src/main.ts` to the other directory with a script. There are some other solutions, but it depends on your environment, so it probably would be better to discuss it in Discord.

**Comment by ark120202** [comment]
> I think the best way to solve it would be providing a new option that would mirror LUA_PATH in runtime.

After testing how `rootDirs` works I think it fits that role quite well

**Comment by lolleko** [comment]
Closed via #1319
